torchvision/models/detection/backbone_utils.py

```diff
@@ -41,13 +41,44 @@ def forward(self, x):
         return x
 
 
-def resnet_fpn_backbone(backbone_name, pretrained, norm_layer=misc_nn_ops.FrozenBatchNorm2d):
+def resnet_fpn_backbone(backbone_name, pretrained, norm_layer=misc_nn_ops.FrozenBatchNorm2d, trainable_layers=3):
     backbone = resnet.__dict__[backbone_name](
         pretrained=pretrained,
         norm_layer=norm_layer)
-    # freeze layers
+    """
+    Constructs a specified ResNet backbone with FPN on top of it. Freezes the specified number of layers in the backbone.
+
+    Examples::
+
+        >>> from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
+        >>> backbone = resnet_fpn_backbone('resnet50', pretrained=True, trainable_layers=3)
+        >>> # get some dummy image
+        >>> x = torch.rand(1,3,64,64)
+        >>> # compute the output
+        >>> output = backbone(x)
+        >>> print([(k, v.shape) for k, v in output.items()])
+        >>> # returns
+        >>>   [('0', torch.Size([1, 256, 16, 16])),
+        >>>    ('1', torch.Size([1, 256, 8, 8])),
+        >>>    ('2', torch.Size([1, 256, 4, 4])),
+        >>>    ('3', torch.Size([1, 256, 2, 2])),
+        >>>    ('pool', torch.Size([1, 256, 1, 1]))]
+        
+    Arguments:
+        backbone_name (string): resnet architecture. Possible values are 'ResNet', 'resnet18', 'resnet34', 'resnet50',
+             'resnet101', 'resnet152', 'resnext50_32x4d', 'resnext101_32x8d', 'wide_resnet50_2', 'wide_resnet101_2'
+        norm_layer (torchvision.ops): it is recommended to use the default value. For details visit:
+            (https://github.com/facebookresearch/maskrcnn-benchmark/issues/267)
+        pretrained (bool): If True, returns a model with backbone pre-trained on Imagenet
+        trainable_layers (int): number of trainable (not frozen) resnet layers starting from final block.
+            Valid values are between 0 and 5, with 5 meaning all backbone layers are trainable.
+    """
+    # select layers that wont be frozen
+    assert trainable_layers<=5 and trainable_layers >=0
+    layers_to_train = ['layer4', 'layer3', 'layer2', 'layer1', 'conv1'][:trainable_layers]
```

Comment:
`conv1` is part of every `Bottleneck` and `BasicBlock`, see https://github.com/pytorch/vision/blob/bd27e942683099dc9aa3d13600c2a9272108eed9/torchvision/models/resnet.py#L91
so I'm not sure that the `conv1` part in here has the expected behavior (it might have, but can be by accident).

Reply:
Sure, I will add test cases and update the PR.

Regarding the conv1 part, I understand that it looks hacky and is not intuitive. But it works it this case because of the way freezing is done. 

Because of the order of strings in layers_to_train 'conv1' is only used when trainable_layers is set to 5 and the entire network should be unfrozen. So even if 'conv1' finds matches in other blocks, its okay.

The fully qualified layer names that the resnet_fpn_backbone observes for res50 are:

conv1.weight
layer1.0.conv1.weight
layer1.0.conv2.weight
layer1.0.conv3.weight
layer1.0.downsample.0.weight
layer1.1.conv1.weight
layer1.1.conv2.weight
layer1.1.conv3.weight
layer1.2.conv1.weight
layer1.2.conv2.weight
layer1.2.conv3.weight
layer2.0.conv1.weight
layer2.0.conv2.weight
layer2.0.conv3.weight
layer2.0.downsample.0.weight
layer2.1.conv1.weight
layer2.1.conv2.weight
layer2.1.conv3.weight
layer2.2.conv1.weight
layer2.2.conv2.weight
layer2.2.conv3.weight
layer2.3.conv1.weight
layer2.3.conv2.weight
layer2.3.conv3.weight
layer3.0.conv1.weight
layer3.0.conv2.weight
layer3.0.conv3.weight
layer3.0.downsample.0.weight
layer3.1.conv1.weight
layer3.1.conv2.weight
layer3.1.conv3.weight
layer3.2.conv1.weight
layer3.2.conv2.weight
layer3.2.conv3.weight
layer3.3.conv1.weight
layer3.3.conv2.weight
layer3.3.conv3.weight
layer3.4.conv1.weight
layer3.4.conv2.weight
layer3.4.conv3.weight
layer3.5.conv1.weight
layer3.5.conv2.weight
layer3.5.conv3.weight
layer4.0.conv1.weight
layer4.0.conv2.weight
layer4.0.conv3.weight
layer4.0.downsample.0.weight
layer4.1.conv1.weight
layer4.1.conv2.weight
layer4.1.conv3.weight
layer4.2.conv1.weight
layer4.2.conv2.weight
layer4.2.conv3.weight
fc.weight
fc.bias

Reply:
One thing I can do to make this more intuitive is to replace :

if all([layer not in name for layer in layers_to_train]):

with 

if all([not name.startswith(layer) for layer in layers_to_train]):

this will make sure that only clean matches are found. What are your thoughts on this ?

Reply:
Here is a [notebook](https://github.com/muaz-urwa/temp_files/blob/master/Faster%20rcnn%20initialization%20behavior%20test.ipynb) which tests the desired functionality.

I will add the test cases based on it into test_models_detection_utils.py

Reply:
> One thing I can do to make this more intuitive is to replace :
> ```python
> if all([layer not in name for layer in layers_to_train]):
> ```
> with
> ```python
> if all([not name.startswith(layer) for layer in layers_to_train]):
> ```
> this will make sure that only clean matches are found. What are your thoughts on this

sure, looks good to me!

Reply:
Test case has been added.